src/main/java/io/confluent/idesidecar/websocket/resources/WebsocketEndpoint.java

```diff
@@ -177,66 +136,39 @@ public void onOpen(Session session) throws IOException {
   public void onMessage(String messageString, Session senderSession) {
     var workspaceSession = sessions.get(senderSession);
     if (workspaceSession == null) {
-      sendErrorAndCloseSession(
-          senderSession,
-          null,
-          "Odd! Received message from unregistered session %s. Closing session.",
-          senderSession.getId()
-      );
-      return;
-    }
+      // Expect to find them in purgatory, and that this is a HELLO_WORKSPACE message
+      if (!purgatorySessions.remove(senderSession)) {
```

Comment:
This line tries to remove the session from purgatory, but if it is successful don't we need to add it to the tracked sessions? It seems like this is done in `handleHelloMessage(...)`, but what happens if that fails? Maybe the removal from purgatory set should be done in the same place we're adding the session to the other map, so that they're always done together (without anything else between these two steps that might fail). Consider being defensive, and removing the session from purgatory _after_ it's added to the other map.

Maybe change this line to:
```suggestion
      if (!purgatorySessions.contains(senderSession)) {
```

Another way to simplify this is to not have separate collections for purgatory and validated workspace sessions, and instead have _one_ map that tracks both states and makes it easier to transition the state of any session. For example, maybe consider separating the logic into a separate `WebsocketSessions` class that encapsulates the map, the state management, and maybe even the `broadcast` logic that could stream over all but filter and skip any sessions that are in purgatory. That class could be an `@ApplicationScoped` bean and `@Injected` into the `WebsocketEndpoints` class, and that would also be a good spot to add in the future the scheduled method that purges sessions with expired TTL from the purgatory state. The public interface of this bean should hide all of the details and would only have the methods needed outside the class, such as getting the `WorkspaceSession` object for either a workspace ID or `Session` object (whichever is needed).

There are other benefits of this approach. 
1. Other classes/beans could broadcast events to sessions by using the `WebsocketSessions` bean, and then leave `WebsocketEndpoints` to be just that -- the endpoints and methods called by the Quarkus framework due to client calls.
2. The `WebsocketSessions` class and the logic we choose to encapsulate is much more easily unit tested, as it's more narrowly scoped than `WebsocketEndpoints`. 

Also, consider a `WorkspaceId` record type, so that the signatures of collections and methods are more strongly typed than with just `String`.

Reply:
Going with your provided revisions of wholistic map of not-yet-hello / post-hello workspace sessions, but directly in the websocket route resource.